### PR TITLE
Add endpoint PUT /provider

### PR DESCRIPTION
**Pull request by fredx30**
This adds a code path for PUT for /api/provider

This goes together with a few other PRs:
- The implementable interface https://github.com/iver-wharf/wharf-api-client-go/pull/8
- The PUT /api/token https://github.com/iver-wharf/wharf-api/pull/26
- (This) The PUT /api/provider https://github.com/iver-wharf/wharf-api/pull/28
- The gtihub-provider changes that require the rest https://github.com/iver-wharf/wharf-provider-github/pull/10

### CODE REVIEW

File: provider.go [hunk 1]
```diff
@@ -185,3 +185,31 @@ func (m ProviderModule) postProviderHandler(c *gin.Context) {
...
+// @failure 401 {object} problem.Response "Unauthorized or missing jwt token"
+// @failure 502 {object} problem.Response "Database is unreachable"
+// @router /provider [put]
+func (m TokenModule) putProviderHandler(c *gin.Context) {
+	var provider Provider
+	if err := c.ShouldBindJSON(&provider); err != nil {
+		problem.WriteInvalidBindError(c, err, "One or more parameters failed to parse when reading the request body.")
+		return
+	}
+	var placedProvider Provider
```
Comment (jilleJr) [inline provider.go:207]: What's this verb "placed"? Uncommon term for variable nomenclature

Suggest renaming the variables to something like

- `provider` -> `inputProvider`
- `placedProvider` -> `dbProvider`

File: provider.go [hunk 2]
```diff
@@ -185,3 +185,31 @@ func (m ProviderModule) postProviderHandler(c *gin.Context) {
 
 	c.JSON(http.StatusCreated, provider)
 }
+
+// putProviderHandler godoc
+// @summary Put provider in database.
+// @description Put provider in database.
```
Comment (jilleJr) [inline provider.go:191]: For those uninitiated by REST verbs, maybe it'd be better to change this to something like:

```suggestion
// @description Creates a new provider if a match is not found.
```

File: provider.go [hunk 3]
```diff
@@ -185,3 +185,31 @@ func (m ProviderModule) postProviderHandler(c *gin.Context) {
...
+// @failure 502 {object} problem.Response "Database is unreachable"
+// @router /provider [put]
+func (m TokenModule) putProviderHandler(c *gin.Context) {
+	var provider Provider
+	if err := c.ShouldBindJSON(&provider); err != nil {
+		problem.WriteInvalidBindError(c, err, "One or more parameters failed to parse when reading the request body.")
+		return
+	}
+	var placedProvider Provider
+	if err := m.Database.Where(token).FirstOrCreate(&placedProvider).Error; err != nil {
```
Comment (jilleJr) [inline provider.go:208]: `.Where(token)` seems out of place here

Maybe it should be like this?

```suggestion
	if err := m.Database.Where(provider).FirstOrCreate(&placedProvider).Error; err != nil {
```

**Review by jilleJr — CHANGES_REQUESTED** [review]

File: provider.go (hunk 1 above)
Comment (fredx30) [reply provider.go:207]: Fair point i think the idea was to use the past tense of PUT. Implementing `provider -> inputProvider` now.

File: provider.go (hunk 3 above)
Comment (fredx30) [reply provider.go]: Yeah that would have been awkward. Thanks!

File: provider.go [hunk 4]
```diff
@@ -198,14 +198,14 @@ func (m ProviderModule) postProviderHandler(c *gin.Context) {
 // @failure 401 {object} problem.Response "Unauthorized or missing jwt token"
 // @failure 502 {object} problem.Response "Database is unreachable"
 // @router /provider [put]
-func (m TokenModule) putProviderHandler(c *gin.Context) {
+func (m ProviderModule) putProviderHandler(c *gin.Context) {
```
Comment (fredx30) [inline provider.go:201]: Also found and fixed this one.

File: CHANGELOG.md [hunk 5]
```diff
@@ -12,7 +12,9 @@ This project tries to follow [SemVer 2.0.0](https://semver.org/).
 	https://changelog.md/
 -->
 
-## v4.0.1 (WIP)
+## v4.1.0 (WIP)
+
+- Add endpoint PUT `PUT /provider`. (#28)
```
Comment (jilleJr) [inline CHANGELOG.md:17]: nit: Double wording of "PUT". Also, not a requirement, but would be nice to say a little what the new endpoint does.

```suggestion
- Added endpoint `PUT /provider` as an idempotent way of creating a
  provider. (#28)
```

**Review by jilleJr — APPROVED** [review]
YetAnotherNitpickInTheChangelog, but the endpoint looks flawless!

File: provider.go [hunk 6]
```diff
@@ -185,3 +185,31 @@ func (m ProviderModule) postProviderHandler(c *gin.Context) {
...
+// @tags provider
+// @accept json
+// @produce json
+// @param provider body Provider _ "provider object"
+// @success 200 {object} Provider
+// @failure 400 {object} problem.Response "Bad request"
+// @failure 401 {object} problem.Response "Unauthorized or missing jwt token"
+// @failure 502 {object} problem.Response "Database is unreachable"
+// @router /provider [put]
+func (m ProviderModule) putProviderHandler(c *gin.Context) {
```
Comment (jilleJr) [inline provider.go:201]: Forgotten to add this method to the `ProviderModule.Register` method

**Review by jilleJr — CHANGES_REQUESTED** [review]
I forgot about this detail :) Lucky that your other PR (#26) brought this up

File: provider.go (hunk 6 above)
Comment (fredx30) [reply provider.go:202]: Its in!
https://github.com/iver-wharf/wharf-api/blame/63c2c4986a42f58394f8db8729265fe5ba154a6d/provider.go#L42

**Review by jilleJr — APPROVED** [review]
LGTM!

**Review by iverestefans — APPROVED** [review]

**Pull request merged**